### PR TITLE
Adding commands to jump to first/last row/column in tables

### LINKED ISSUE #957: Commands to facilitate navigation on the tables
**Reported by EltonLopes on 2010-10-02 18:00**
In a table, would like the following commands to facilitate navigation on the tables and also out of them:

Control + Alt + Home: Go to the first table cell
Control + Alt + End: Go to the last cell of the table
Control + Alt + Page Up: Jump Up cell x in the same column
Control + Alt + Page Down: Jumps x cells down in the same column

### PR DESCRIPTION
**Pull request by mltony**
### Link to issue number:
fixes #957

### Summary of the issue:
Introduce commands to jump to first/last row/column in current table.
### Description of how this pull request fixes the issue:
Implements new commands in class DocumentWithTableNavigation.
### Testing strategy:
* Tested in Chrome, Firefox, Edge.
* Tested in HTML tables with merged cells provided in https://github.com/nvaccess/nvda/pull/13345#issuecomment-1046376381
* Also tested on MSWord table from previous comment in both browse and focus mode.
* Tested in HTML table with missing cells.
### Known issues with pull request:
Will cause merge conflicts with #13345 - will be happy to resolve them once it lands.
### Change log entries:
New features
* New table navigation commands: Control+Alt+Home/End to jump to first/last column; Control+Alt+PageUp/PageDown to jump to first/last row.
Changes
Bug fixes
For Developers

### Code Review Checklist:

- [x] Pull Request description:
  - description is up to date
  - change log entries
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] API is compatible with existing add-ons.
- [x] Documentation:
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English

### CODE REVIEW

**Comment by mltony** [comment]
cc: @feerrenrut, @seanbudd

File: source/documentBase.py
Comment (lukaszgo1) [inline source/documentBase.py:94]: For new code types for parameters and return values should be specified using Python type annotations if possible. The same goes for all docstrings below - I'm not going to comment on all of them.

File: source/documentBase.py
Comment (lukaszgo1) [inline source/documentBase.py:163]: While this should not block this PR ideally both axes and directions  of movement should be specified as an enum.

File: source/documentBase.py
Comment (lukaszgo1) [inline source/documentBase.py:169]: A comment here explaining how you come with the 1000 as a right value would be nice.

File: source/documentBase.py
Comment (lukaszgo1) [inline source/documentBase.py:294]: For new scripts script decorator should be used to set script's metadata. Obviously for `documentBase` this cannot be done at present as it would reintroduce #12399, but perhaps an issue should be created to at least avoid this problem being forgotten.

File: source/virtualBuffers/gecko_ia2.py
Comment (lukaszgo1) [inline source/virtualBuffers/gecko_ia2.py:520]: I don't see how `OSError` may be raised by this code. Could you elaborate? At a guess this should be a `COMError`.

**Review by lukaszgo1 — CHANGES_REQUESTED** [review]
Please add  the newly introduced gestures to the user guide.

File: source/documentBase.py
Comment (mltony) [reply source/documentBase.py:94]: Fixed that. The only thing I couldn't annotate is table_id as it is a union. This union is properly defined in #13345 - will annotate once it lands.

File: source/documentBase.py
Comment (mltony) [reply source/documentBase.py:163]: Already implemented in #13345.

File: source/documentBase.py
Comment (mltony) [reply source/documentBase.py:169]: This is pretty much an arbitrary large number. I added a pretty obvious comment. There is no science behind it.

File: source/documentBase.py
Comment (mltony) [inline source/documentBase.py:295]: I don't have the background on why this would be impossible, so I don't feel I can create a meaningful issue.

File: source/virtualBuffers/gecko_ia2.py
Comment (mltony) [reply source/virtualBuffers/gecko_ia2.py:520]: Good catch! Deleted OSError.

**Comment by mltony** [comment]
@lukaszgo1,  Addressed your comments.

**Comment by seanbudd** [comment]
The build is failing as linting needs to be fixed.

**Comment by Adriani90** [comment]
@mltony can the commands be reassigned by the user? Me personally for example I would prefer to use ctrl+alt+nvda+arrow keys to jump to last column or row. The thing is that my laptop keyboard does not have home, end, page up and page down keys, so I had to replace some f keys with sharp keys to acomplish this function. But is quite unconfortable to use them very often.

File: user_docs/en/changes.t2t
Comment (CyrilleB79) [inline user_docs/en/changes.t2t:26]: Please remove the change on this file. To avoid dealing with many merge conflicts on the change log, NVAccess prefers to update this file themselves. Once your PR is accepted, they will look at the change log section in the initial description of the PR and they will copy/paste its content to this file.

**Review by CyrilleB79 — COMMENTED** [review]
Hi Tony

Really happy to see this PR arriving in core!

Some comments here:
* You should test (and implement if needed) the same type of commands in list views where table navigation commands are also available.
  If this is not possible for any reason, you should document it and explain why in this PR's initial description.
  Note: IMO, if this is not possible for list views, it would already be a great improvement to merge this PR only for real tables.. But it must be clearly documented.
* Testing: For Word, did you test with and without UIA enabled? Please update the test section in the initial description accordingly.

Thanks!

**Review by Qchristensen — APPROVED** [review]
User Guide change is good, thanks - good work!

**Review by michaelDCurran — CHANGES_REQUESTED** [review]
It is possible to write code to get the dimensions of the table without having to resort to hit testing to find the limits.
From any given textInfo position, you can expand to character, getTextWithFields,  walk up the ancestor controlFields until reaching the inner most table controlField, and then fetch table-rowcount and table-columncount attributes.
Refer to `_getTableCellCoords` for a similar example of where attributes are fetched off a table controlField.

**Comment by mltony** [comment]
@seanbudd, fixed lint and build.
@Adriani90, no problem to reassign keystrokes.
@CyrilleB79, codepath is entirely different for list views (it is implemented in RowWithFakeNavigation class IIRC), so I'd rather separate that into another PR. Also, for word, I tried turning UIA on and off for word in NVDA settings, and it seems to be still  using UIA even when turned off. But now with Michael's suggestion I feel this is irrelevant since we don't need to interact with low-level IA2/UIA API.
@michaelDCurran, thanks, changed code according to your suggestion. Also refactored a bit to avoid code reduplication.
The bot says system tests failed, but from what I can see, it appears that only 1 test is broken and it's probably transient and does not appear to be related:
```
[00:24:52] Starts from desktop shortcut :: Ensure that NVDA can start from de... | FAIL |
[00:24:52] Connection to remote server broken: [WinError 10061] No connection could be made because the target machine actively refused it
```

**Comment by mltony** [comment]
Also PR #13470 provides same keystrokes in list views and tree views.

**Review by michaelDCurran — APPROVED** [review]

**Comment by seanbudd** [comment]
@mltony - can you resolve the merge conflicts as you mentioned in the PR description. I've just merged #13345

**Comment by mltony** [comment]
@seanbudd, resolved conflicts!

File: source/documentBase.py
Comment (seanbudd) [inline source/documentBase.py:239]:
```suggestion
			movement: _Movement,
			axis: _Axis
```

File: source/documentBase.py
Comment (seanbudd) [inline source/documentBase.py:255]:
```suggestion
		@param movement: the direction ("first"/"next" or "last"/"previous")
```

**Review by seanbudd — COMMENTED** [review]
LGTM, just some changes on documentation / types

**Comment by mltony** [comment]
@seanbudd, fixed and rebased.

**Review by seanbudd — APPROVED** [review]
thanks @mltony.

please avoid rebases in future for minor changes if possible. After a rebase, GitHub no longer allows you to review pushed changes since your last review. This means the entire PR must be re-reviewed, instead of the changes.

**Pull request merged**
